homeassistant/components/stream/__init__.py

```diff
@@ -200,6 +209,14 @@ def _stop(self):
             _LOGGER.info("Stopped stream: %s", self.source)
 
 
+def _get_stream_by_source(hass, stream_source):
```

Comment:
Why change the data type from a list to a dict if we have to end up looking it up?

Reply:
My thinking was that stream url can no longer be the primary key since we want to support changing the url.

I thought of the new behavior as similar to how the access token check is done: on every request for a stream sequence it iterates over the streams to find the one with the matching access token.

However, it's not the only approach we could do.  Maybe also:
- change how camera invokes record, so that the lookup of existing worker sn't by stream url but by using the same stream object (considered this but was keeping diff small)
- support add/remove of the dict when the url changes
- use a different key such as camera entity id

One thing I don't have intuition for is how often the resuse of existing stream happens by calls to stream record vs calls to camera record, so I leaned towards preserving behavior as much as possible.

Reply:
I think the access token check is harder to avoid since we end up resetting the tokens periodically. For the streams itself I think it's easy to skip that iterating check by storing the streams by some other key, but as you said we will need to change the `async_handle_record_service` call to use that other key as well.

Reply:
Yeah, my assumption is that what is here is better than a backwards incompatible service API call change for camera.  Even if we added an additional optional key for usage by camera, record seems like it needs to work with a raw rtsp stream url without that key, so it would still need to iterate through everything and match the rtsp stream url anyway.

Reply:
I don't think there is a problem changing the service call - it looks like it's just an internal service call between camera and stream. Instead of camera calling the service with `CONF_STREAM_SOURCE: source` we could just have camera call the service with the new key, and then on the stream side it can look up the stream directly with that key.

Reply:
Are you suggesting it would be OK to remove `stream_source` from the API?

I assumed that since `stream.record` shows up in the home assistant user interface that users can build automations, etc,  to record arbitrary rtsp stream urls without a camera. (I just tried this out and it).  Just to clarify: you're saying we can break that functionality because it was never intended to be used outside of `camera`?

That is, i understand how to make the change, but don't have enough context to know that it is ok to break that feature.

Reply:
You're right, I didn't realize it was exposed there and documented as a separate service. I would be an advocate of removing that service since I don't see a use for it on its own, but that would be a breaking change and we would need an OK from some higher ups. Let's see what @hunterjm says.

Reply:
OK so, given the other comment thread, we'll proceed with this breaking change.  So, then what I'll do is change the service to accept a `stream_id` of a previously created stream.  This means that `record` can no longer be invoked on a stream that does not already exist, and so the camera component will make sure the stream exists before calling record.  Thanks.

Reply:
Just to be clear, I think you mean that these changes mean that stream's `record` can no longer be invoked independently on an arbitrary url and will only called by camera, and since camera is the caller it will make sure the stream is created before calling record? I think that makes sense

Reply:
Yes, ok.  Given the discussion on the main comment thread, i've changed the interface to record to take a `stream_id` rather than a `stream_source` which addresses this.  Streams can now be looked up by that id in the `ATTR_STREAMS` dict again.

Reply:
Based on the other thread, the record service was removed entirely.